src/GridFS/Exception/StreamException.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace MongoDB\GridFS\Exception;
+
+use MongoDB\Exception\RuntimeException;
+use function MongoDB\BSON\fromPHP;
+use function MongoDB\BSON\toJSON;
+use function sprintf;
+use function stream_get_meta_data;
+
+class StreamException extends RuntimeException
+{
+    /** @param resource $destination */
+    public static function downloadFromFilenameFailed(string $filename, $destination) : self
+    {
+        $metadata = stream_get_meta_data($destination);
+
+        return new static(sprintf('Downloading GridFS file "%s" to resource "%s" failed: stream error.', $filename, $metadata['uri']));
```

Comment:
Apologies for digging into this further, but some new thoughts on this.

Does "stream error" add anything useful here, given that all StreamException variants use the same suffix? Perhaps remove it to keep the message shorter. Since we're suppressing the PHP error, it occurs to me that we really don't have any contextual message to report other than "something went wrong" -- and even if we didn't suppress the error we aren't in a position to capture it for inclusion here.

Neither the filename nor ID really uniquely identify the file since we don't know the bucket namespace -- and even a filename isn't unique within a bucket namespace. Would it make sense to also use the `$metadata['uri']` of the GridFS download stream here? By this point, our Bucket method (i.e. `openDownloadStream`, `openDownloadStreamByName`) will already have created a file path (i.e. `createPathForFile`) and that should uniquely identify the GridFS file being downloaded (bucket namespace appended with ID). It conveniently converts ObjectId identifiers to a string (only deferring to extended JSON if necessary because a `__toString()` doesn't exist).

We can still include the user-specified ID or filename as well, as in the following examples:

    Downloading file from "<gridfs-uri>" to "<destination-uri>" failed. GridFS identifier: <id>
    Downloading file from "<gridfs-uri>" to "<destination-uri>" failed. GridFS filename: <filename>

More on `openUploadStream` below.

Reply:
Good point. Updated and added tests